pkg/controllers/avuploader/controller.go

```diff
@@ -0,0 +1,175 @@
+		return reconcile.Result{}, nil
+	}
+
+	request := constructAvsRequest(*availabilityCollection)
+
+	err := doAvsRequest(request, c.Config().AvailabilityMonitoring.AvailabilityServiceConfiguration.Url, c.Config().AvailabilityMonitoring.AvailabilityServiceConfiguration.ApiKey)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	availabilityCollection.Status.LastReported = availabilityCollection.Status.LastRun
+
+	//write to status
+	if err := c.Client().Status().Update(ctx, availabilityCollection); err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to update availability status: %w", err)
+	}
+
+	return reconcile.Result{}, nil
+
+}
+
+func constructAvsRequest(availabilityCollection lssv1alpha1.AvailabilityCollection) AvsRequest {
+	//Fill failedInstances with all failed. A failed instance will create an instance outage. If this instance is not in the array anymore, instance outage is resolved
+	// Overall status is derrived if len(failedInstances) > 0
+	failedInstances := []AvsInstance{}
+	for _, instanceStatus := range availabilityCollection.Status.Instances {
+		if instanceStatus.Status == FAILED {
+			avsInstance := AvsInstance{
+				InstanceId:   instanceStatus.Name,
+				Name:         instanceStatus.Name,
+				Status:       AVS_STATUS_DOWN,
+				OutageReason: instanceStatus.FailedReason,
+			}
+			failedInstances = append(failedInstances, avsInstance)
+		}
+	}
+	if availabilityCollection.Status.Self.Status == FAILED {
+		avsInstance := AvsInstance{
+			InstanceId:   "Self",
+			Name:         "Self",
+			Status:       AVS_STATUS_DOWN,
+			OutageReason: availabilityCollection.Status.Self.FailedReason,
+		}
+		failedInstances = append(failedInstances, avsInstance)
+	}
+
+	status := AVS_STATUS_UP
+	outageReson := ""
+	if len(failedInstances) > 0 {
+		status = AVS_STATUS_DOWN
+		//include self landscaper (--> +1 )
+		totalNumberOfMonitoredLandscapers := len(availabilityCollection.Status.Instances) + 1
+		outageReson = fmt.Sprintf("%d/%d monitored landscaper down", len(failedInstances), totalNumberOfMonitoredLandscapers)
+	}
+
+	request := AvsRequest{
+		Timestamp:          availabilityCollection.Status.LastRun.Unix(),
+		ResponseTime:       0,
+		ResponseStatusCode: 200,
+		Instances:          failedInstances,
+		Status:             status,
+		OutageReason:       outageReson,
+	}
+	return request
+}
+
+func doAvsRequest(request AvsRequest, url string, apiKey string) error {
+	body, err := json.Marshal(request)
+	if err != nil {
+		return fmt.Errorf("avs upload post payload json marhsal failed: %w", err)
+	}
+	client := &http.Client{}
+	avsreq, err := http.NewRequest("POST", url, bytes.NewBuffer(body))
+	if err != nil {
+		return fmt.Errorf("avs upload request build failed: %w", err)
+	}
+	avsreq.Header.Add("Api-Key", apiKey)
+	avsreq.Header.Add("Content-Type", "application/json")
+	avsreq.Header.Add("accept", "application/json")
+	resp, err := client.Do(avsreq)
```

Comment:
This seems to be a blocking call?
Could we handle the call asynchronously? If not, shall or can we specify a timeout to avoid deadlocks?

Reply:
Do you know if the controller-runtime framework handles the different controllers in goroutines so we would not block other reconciler?
If yes, I would go for timeout, since the AVS uploader can not do something in parallel, so it has to block anyway.

Reply:
Yes, each controller has a seperate worker pool (the default size is one): https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller#Options
This is something we could make configurable in the future, like the landscaper does.

So yes, I think a timeout would be sufficient.